### PR TITLE
WIP: Ephemeral connections

### PR DESCRIPTION
**Pull request by pegahgh**
This PR investigates effect of adding ephemeral connection  and no layer wise pre-training in TDNN (issue #988 ).

### CODE REVIEW

File: src/nnet3bin/nnet3-am-copy.cc [hunk 1]
```diff
@@ -143,6 +144,9 @@ int main(int argc, char *argv[]) {
 
     if (scale != 1.0)
       ScaleNnet(scale, &(am_nnet.GetNnet()));
+    
+    if (dropout > 0)
+      SetDropoutProportion(dropout, &((am_nnet.GetNnet())));
```
Comment (GaofengCheng) [inline src/nnet3bin/nnet3-am-copy.cc]: delete the extra parentheses

File: egs/wsj/s5/steps/nnet3/tdnn/make_configs.py [hunk 2]
```diff
@@ -116,6 +116,22 @@ def GetArgs():
     parser.add_argument("--use-presoftmax-prior-scale", type=str, action=nnet3_train_lib.StrToBoolAction,
                         help="if true, a presoftmax-prior-scale is added",
                         choices=['true', 'false'], default = True)
+    # arguments for ephemeral connection
+    parser.add_argument("--add-ephemeral-connection", type=str, action=nnet3_train_lib.StrToBoolAction,
+                        help="if true, a ephemeral connection with dropout connection is added to all layers.",
+                        choices=['true', 'false'], default = False)
+    parser.add_argument("--num-skips-for-ephemeral", type=int,
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/tdnn/make_configs.py]: --num-layer-skips-for-ephemeral may be better, in BLSTM we may intro ephemeral in time axis, we should prevent ambiguity

File: egs/wsj/s5/steps/nnet3/tdnn/make_configs.py [hunk 3]
```diff
@@ -458,10 +458,30 @@ def MakeConfigs(config_dir, splice_indexes_string,
...
-                                                            self_repair_scale = self_repair_scale,
-                                                            norm_target_rms = 1.0 if i < num_hidden_layers -1 else final_layer_normalize_target)
+                if add_ephemeral_connection and i > 0:
+                    dropout_name = "Tdnn_{0}".format(i-2)
+                    if i == 1:
+                      dropout_name = "Tdnn_init"
+
+                    if (nonlin_output_dim_final == nonlin_output_dim_init) and i > num_skips_for_ephemeral:
+                        prev_layer_output = nodes.AddAffRelNormWithDirectEphemeralLayer(config_lines, 
+                                                                    "Tdnn_{0}".format(i), prev_layer_output,
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/tdnn/make_configs.py]: make sure "Tdnn_{0}" lines up with config_lines

File: egs/wsj/s5/steps/nnet3/tdnn/make_configs.py [hunk 4]
```diff
@@ -458,10 +458,30 @@ def MakeConfigs(config_dir, splice_indexes_string,
...
+                                                                    dropout_name, dropout_layer_input,
+                                                                    nonlin_output_dims[i],
+                                                                    self_repair_scale = self_repair_scale,
+                                                                    norm_target_rms = 1.0 if i < num_hidden_layers -1 else final_layer_normalize_target)
+                    else:
+                        prev_layer_output = nodes.AddAffRelNormWithEphemeralLayer(config_lines, 
+                                                                    "Tdnn_{0}".format(i), prev_layer_output,
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/tdnn/make_configs.py]: "Tdnn_{0}" lines up with config_lines

File: egs/wsj/s5/steps/nnet3/tdnn/make_configs.py [hunk 5]
```diff
@@ -458,10 +458,30 @@ def MakeConfigs(config_dir, splice_indexes_string,
...
+                                                                    dropout_name, dropout_layer_input,
+                                                                    nonlin_output_dims[i],
+                                                                    self_repair_scale = self_repair_scale,
+                                                                    norm_target_rms = 1.0 if i < num_hidden_layers -1 else final_layer_normalize_target)
+                else:
+                    prev_layer_output = nodes.AddAffRelNormLayer(config_lines, "Tdnn_{0}".format(i),
+                                                                prev_layer_output, nonlin_output_dims[i],
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/tdnn/make_configs.py]: "prev_layer_output" line up with "config_lines"

File: egs/wsj/s5/steps/nnet3/tdnn/make_configs.py [hunk 6]
```diff
@@ -477,36 +497,44 @@ def MakeConfigs(config_dir, splice_indexes_string,
...
-        config_lines = {'components':[], 'component-nodes':[]}
+                                   use_presoftmax_prior_scale = use_presoftmax_prior_scale,
+                                   prior_scale_file = prior_scale_file,
+                                   include_log_softmax = include_log_softmax,
+                                   add_final_sigmoid = add_final_sigmoid,
+                                   objective_type = objective_type)
+                if xent_regularize != 0.0:
+                    nodes.AddFinalLayer(config_lines, prev_layer_output, num_targets,
+                                        ng_affine_options = " param-stddev=0 bias-stddev=0 learning-rate-factor={0} ".format(
+                                              0.5 / xent_regularize),
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/tdnn/make_configs.py]: line 509 and 510 could be rearranged

File: egs/wsj/s5/steps/nnet3/tdnn/make_configs.py [hunk 7]
```diff
@@ -477,36 +497,44 @@ def MakeConfigs(config_dir, splice_indexes_string,
...
     # write the files used by other scripts like steps/nnet3/get_egs.sh
     f = open(config_dir + "/vars", "w")
     print('model_left_context=' + str(left_context), file=f)
     print('model_right_context=' + str(right_context), file=f)
-    print('num_hidden_layers=' + str(num_hidden_layers), file=f)
+    if layerwise_pretrain:
+      print('num_hidden_layers=' + str(num_hidden_layers), file=f)
+    else:
+      print('num_hidden_layers=' + str(1), file=f)
+      print('num_real_hidden_layers=' + str(num_hidden_layers), file=f)
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/tdnn/make_configs.py]: In my point of view, 'num_of_layer_configs' and 'num_hidden_layers' sound more natural.

File: egs/wsj/s5/steps/nnet3/components.py [hunk 8]
```diff
@@ -117,6 +117,51 @@ def AddAffRelNormLayer(config_lines, name, input, output_dim, ng_affine_options
     return {'descriptor':  '{0}_renorm'.format(name),
             'dimension': output_dim}
 
+def AddAffRelNormWithEphemeralLayer(config_lines, name, input, ephemeral_name, ephemeral_input, output_dim, ng_affine_options = " bias-stddev=0 ", norm_target_rms = 1.0, self_repair_scale = None):
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/components.py]: Better  not go over 80 characters.

File: egs/wsj/s5/steps/nnet3/train_dnn.py [hunk 9]
```diff
@@ -182,7 +182,7 @@ def GetArgs():
                         help="""If true, remove egs after experiment""")
     parser.add_argument("--cleanup.preserve-model-interval", dest = "preserve_model_interval",
                         type=int, default=100,
-                        help="Determines iterations for which models will be preserved during cleanup. If mod(iter,preserve_model_interval) == 0 model will be preserved.")
+                        help="Determines iterations for which models will be preserved during cleanup. If iter % preserve_model_interval == 0 model will be preserved.")
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/train_dnn.py]: Better not go over 80 characters.

File: egs/wsj/s5/steps/nnet3/train_dnn.py [hunk 10]
```diff
@@ -555,6 +578,17 @@ def Train(args, run_opts):
     # egs_dir will be updated if there is realignment
     cur_egs_dir=egs_dir
 
+    # If true, compute the dropout proportion for training.
+    # and write dropout proportion.
+    dp_prop = [1.0] * num_iters
+    if add_ephemeral_connection:
+      init_zero_dp_iter = 10 + num_hidden_layers * args.add_layers_period
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/nnet3/train_dnn.py]: Maybe we could add a arg naming "dropout_schedule" to make the changing of dropout rate more easily.

Comment (pegahgh) [inline egs/wsj/s5/steps/nnet3/train_dnn.py]: Sure, adding dropout_schedule  is good too! ComputeDropout computes dropout rate per iteration. 
"num_init_zero_dropout" is also important variable. The whole idea is that we should not do dropout in the beginning of training to let the network use this connection and then increase dropout gradually using "dropout_schedule"! 
We should do more experiments to find the optimum "dropout_schedule".

File: egs/wsj/s5/steps/nnet3/tdnn/make_configs.py (hunk 2 above)
Comment (pegahgh) [inline egs/wsj/s5/steps/nnet3/tdnn/make_configs.py]: wil rename it.

File: egs/wsj/s5/steps/nnet3/tdnn/make_configs.py (hunk 7 above)
Comment (pegahgh) [inline egs/wsj/s5/steps/nnet3/tdnn/make_configs.py]: Sure! I will rename it as num_layers_for_config and num_hidden_layers

File: egs/wsj/s5/steps/nnet3/components.py (hunk 8 above)
Comment (pegahgh) [inline egs/wsj/s5/steps/nnet3/components.py]: Thanks for detailed review, I will fix the issues.

File: egs/wsj/s5/steps/nnet3/train_dnn.py [hunk 11]
```diff
@@ -375,7 +375,28 @@ def TrainOneIteration(dir, iter, srand, egs_dir,
...
-        raw_model_string = "nnet3-am-copy --raw=true --learning-rate={0} {1}/{2}.mdl - |".format(learning_rate, dir, iter)
+        
+        if add_ephemeral_connection:
+          if dropout_proportion == 1.0 and not os.path.exists('{0}/configs/dropout_removed'.format(dir)):
+            # If dropout_proportion is equal to 1.0, 
+            # we replace Tdnn_{0}_relu input from sum(affine, ephemeral_affine) to affine.
+            # and remove orphan nodes and components using edits-config.
+            replace_config_str = '{0}/configs/dropout.config'.format(dir)
+            replace_config_file = open(replace_config_str, 'w')
+            for layer in range(1, num_hidden_layers):
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/train_dnn.py]: Can you explain what the difference between num_hidden_layers and num_real_hidden_layers is?
I don't like the name 'num_real_hidden_layers' much.  What is this used for?

Comment (pegahgh) [inline egs/wsj/s5/steps/nnet3/train_dnn.py]: I have two num_hidden_layers and num_layers_for_config. num_layers_for_config is the number of layer.config files used for layer-wise pretraining.

I can simply also do that by adding layerwise_pretrain in vars, to skip adding layer by layer during training.
Which one do you prefer?

**Comment by danpovey** [comment]
[i commented by email but my comment was lost by github, reposting.]
I think it will be simpler to just write num_hidden_layers and have it be the number of layer.config files needed.  
Inside the python script, make sure the variable names are easy to understand.

**Comment by danpovey** [comment]
I think it will be simpler to just write num_hidden_layers and have it be
the number of layer.config files needed.
Inside the python script, make sure the variable names are easy to
understand.
